daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataLoc.scala

```diff
@@ -33,9 +33,20 @@ import org.apache.daffodil.util.MaybeULong
 import org.apache.daffodil.api.DataLocation
 import org.apache.daffodil.processors.parsers.PState
 
-class DataLoc(val bitPos1b: Long, bitLimit1b: MaybeULong, val isAtEnd: Boolean, eitherStream: Either[DataOutputStream, DataInputStream],
+class DataLoc(
+  val bitPos1b: Long,
+  bitLimit1b: MaybeULong,
+  eitherStream: Either[DataOutputStream, DataInputStream],
   val maybeERD: Maybe[ElementRuntimeData]) extends DataLocation {
 
+  @deprecated("Use bitPos1b to compare with expected position (possibly bitLimit1b).", "3.1.0")
+  override def isAtEnd =
+    eitherStream match {
+      case Right(isdis: InputSourceDataInputStream) => isdis.isAtEnd
```

Comment:
I"m not sure this is worth worrying about, but this potentially has different behavior if you do something like this:
```scala
val res1 = dp.parse(isdis, ...)
val loc1 = res.location()
val res2 = dp.parse(isdis, ...)
val loc2 = res.location()
println(loc1.isAtEnd)
println(loc2.isAtEnd)
```
So if our InputSourceDataInputStream hit an EOD at the end of the second parse (i.e. our stream had two messages), then both ``loc1.isAtEnd`` and ``loc2.isAtEnd`` would be true. But only loc2.isAtEnd should be true. So future parse calls can change the isAtEnd state of previous ParseResults. Maybe this isn't a big deal since I doubt anyone will check isAtEnd of a previous parse location, but maybe it's worth considering? We at least probably want a backwards compatability note somewhere about this if we do keep this.

Could maybe do some logic where DataLoc also captures the bitPos of the InputSourceDataInputStream when it's created. If the data location has changed, then isAtEnd returns false, otherwise it returns isids.isAtEnd? Seems complicated for such a small edge case.

Another alternative would maybe be to have a tunable that complete disables use of isAtEnd in the DataLoc. Depending on the value of that tunable, DataLoc either calls isdis.isAtEnd and stores it (and blocking on TCP connections), or it just always returns false. Maybe call it something like "requireAccurateIsAtEnd" or something. People using TCP that aren't going to use isAtEnd can set this to false. And we default it to true for perfect backwards compatibility. Again, might be extra complication for a small edge case, so not sure it's worth it.

Just want to through out some ideas an see if it sparks any thoughts.

Agree that we definitely need to deprecate isAtEnd and its use though.

Reply:
Ultimately, a ParseResult should be an immutable object, excepting the InputSourceDataInputStream within it, which is obviously mutable. 
We can't answer isAtEnd without potentially mutating the ISDIS, so that can't be part of an immutable ParseResult. 
So really the problem is loc.isAtEnd shouldn't be on the loc object, because we want the loc object to be immutable. That's a big reason to deprecate isAtEnd. The bit pos and bit limit information on a DataLoc is immutable. 

In the latest commit, hasData() is on the DataInputStream, and there is no equivalent on DataLocation/DataLoc. 

I think adding a tunable here is overkill. Deprecating isAtEnd on loc, is easy to replace it with getting the data input stream and calling hasData().

Reply:
All fair. Sounds good to me.